[FEAT] Setting 관련 View와 Feature(Store) 구현

importantsgit opened the pull request:
## 변경 사항 설명
- Setting(Profile) Flow 내 View와 Feature을 개발했습니다.

## 변경된 사항을 작성 해주세요
- 설정(프로필) 페이지 View, Feature 추가
- 피드백 페이지 View, Feature 추가
- 알람 설정 페이지 View, Feature 추가
- 전체 한 일 페이지 View, Feauture 추가

## 변경 유형
<!-- 해당하는 항목에 x 표시를 해주세요. -->
- [ ] 버그 수정
- [x] 새로운 기능
- [ ] 코드 스타일 업데이트 (포맷팅, 변수명 등)
- [ ] 리팩토링 (기능 변경 없음)
- [ ] 문서 내용 변경
- [ ] 기타 (설명해주세요):

## 체크리스트
<!-- 완료한 항목에 x 표시를 해주세요. -->
- [ ] 코드가 프로젝트의 스타일 가이드를 준수합니다.
- [ ] 변경 사항에 대한 테스트를 추가했습니다.
- [ ] 모든 새로운 기능과 변경 사항이 문서화되었습니다.
- [ ] PR이 적절한 크기로 나누어졌습니다.

## 추가 정보
<!-- 리뷰어가 알아야 할 추가 정보가 있다면 여기에 적어주세요. -->
- 개선해야 할 사항
   > GUI를 DesignSystem에 맞게 수정해야 합니다.
   > Feature의 reduce 부분을 수정해야 합니다.
   > 별도의 화면 이동 로직이 포함되지 않았습니다.

importantsgit committed:
[FEAT] Setting 관련 View와 Feature(Store) 구현

수정해야 할 사항
- 현재 완벽하게 UI를 구성하지 않았습니다.
- 추후 프로젝트 디자인 시스템에 맞춰 작업하겠습니다.

Marcorable commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/Feedback/FeedbackView.swift:
```diff
@@ -0,0 +1,122 @@
+        
+        VStack {
+            NavigationBarView(title: "피드백 주기")
+            
```
혹시 View 간의 간격을 두신 이유가 있을까요?

Marcorable commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/Feedback/FeedbackView.swift:
```diff
@@ -0,0 +1,122 @@
+
+fileprivate struct FeedbackInputView: View {
+    
+    @Bindable var store: StoreOf<FeedbackFeature>
```
하위 뷰 컴포넌트 단위에서는 설계해주신 NavigationBarView 처럼 외부에서 값을 주입 받아서 구성하면 어떨까요?

컴포넌트 단위의 디자인을 구성할 때 프리뷰를 따로 구성 할 수도 있는데
storeOf 바인딩 받게되면 프리뷰 구성마다 초기화 해주어야 하기도 하고
원하는 뷰의 상태를 구성하기가 까다롭기도 하더라구요!

Marcorable commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/SettingView.swift:
```diff
@@ -0,0 +1,223 @@
+                Text("지금까지 한 일 보기")
+                    .font(.system(size: 14, weight: .bold))
+                    .frame(width: 260, height: 42)
+            }
```
혹시 이 버튼은 가로 세로 크기는 기기에 상관없이 동일한 방향으로 구현하신 건가요?

이 버튼에 의도를 모르는 상태라서 조심스럽지만..
화면에 중앙에서 어느정도 비중을 가지는 크기면 좋을것 같이 보이더라구요

SwiftUI 는 UIKit 과 달리 AutoLayout 개념이 없기 때문에
다른 방식의 Layout을 구성하면 좋은데요

제가 사용하는 방식을 소개해드려 볼까 합니다

```swift
Button {
    store.send(.showTotalTasksButtonTapped)
} label: {
   HStack {
      Spacer()
      Text("지금까지 한 일 보기")
         .font(.system(size: 14, weight: .bold))
         .tint(.white)
         .padding(.vertical) // 고정된 높이 지정가능
         Spacer()
   }
   .background(.black)
   .clipShape(.rect(cornerRadius: 10))
}
.padding(.horizontal, 48)
```

View 개체(컴포넌트)가 상위 화면에서 Layout이 늘어나는 성격을 주어지면 좋더라구요
Vertical / Horizontal / All
위와 같이 구성하면 버튼은 Horizontal 성격을 가지고
개체로 분리한다면 버튼에 주어진 ViewModifier
`.padding(.horizontal, 48)`는 생략하는 형태로 구성 할 수 있을것 같아요

Marcorable commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/SettingView.swift:
```diff
@@ -0,0 +1,223 @@
+    @Bindable var store: StoreOf<SettingFeature>
+    
+    var body: some View {
+        HStack {
+            VStack {
+                Text("지금까지 한 일")
+                
+                Spacer()
+                    .frame(height: 8)
+                
+                Text(String(store.totalTasks))
+                    .font(.system(size: 48, weight: .medium))
+            }
+            .frame(width: 145, height: 118)
+            
+            VStack {
+                Text("하루 평균 한 일")
+                
+                Spacer()
+                    .frame(height: 8)
+                
+                Text(String(store.avgTasksPerDay))
+                    .font(.system(size: 48, weight: .medium))
+            }
+            .frame(width: 145, height: 118)
+        }
```
위에서 언급했던 부분과 유사하게 보여서 다른 스타일을 제안해봅니다.

```swift
HStack {
   Spacer()
   VStack(spacing: 8) {
      Text("지금까지 한 일")
      Text(String(store.totalTasks))
          .font(.system(size: 48, weight: .medium))
   }
   .padding(.vertical)
   Spacer() // HStack(spacing: ...) 으로 고정된 값 대체가능
   VStack(spacing: 8) {
      Text("하루 평균 한 일")
      Text(String(store.avgTasksPerDay))
         .font(.system(size: 48, weight: .medium))
   }
Spacer()
}
```

Stack 에서는 spacing 을 이용해서 고정값을 대체할 수 있을것 같아요!

GeometryReader 와`.frame()` 을 사용하는 경우가 많은데
View 프로토콜의 동작 원리를 이해하고 사용하기보다 강제적으로 컨트롤 하는 느낌이라서
`.frame` 보다는 `.padding` 을 이용하면 AutoLayout을 이용하는 것처럼 더 다양하게 
Layout을 구성할 수 있지 않을까 싶어요!

Marcorable reviewed (commented):
SwiftLint 가 적용 되었을때 컨벤션이 적용이 안된 경우가 있던데 확인한번 부탁드립니다!

importantsgit committed:
STYLE: Lint 스타일 적용

Guboneui commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/Feedback/FeedbackFeature.swift:
```diff
@@ -0,0 +1,70 @@
+
+    // MARK: - body
+
+    var body: some Reducer<State, Action> {
+        Reduce { state, action in
+            switch action {
+            case let .feedbackChanged(feedback):
+                state.feedback = feedback
+
+                print(feedback)
+
+                return .none
+
+            case let .emailChanged(email):
+                state.email = email
+                state.isEmailValid = isValidEmail(email)
+
+                print(email)
+
+                return .none
+
+            case .confirmButtonTapped:
+
+                print("confirmButtonTapped")
+
+                return .none
+            }
```
불필요한 디버깅 구문(print) 는 지워도 될 것 같습니다!

Guboneui commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/Feedback/FeedbackView.swift:
```diff
@@ -0,0 +1,138 @@
+        }
+        .background(isActiveButton ? .black : .white)
+        .clipShape(.capsule)
+
```
```suggestion
```

Guboneui commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/NotificationSetting/NotificationSettingFeature.swift:
```diff
@@ -0,0 +1,74 @@
+    }
+}
+
+extension Date {
```
해당 파일이 아닌 `Date+Extensions` 라는 파일로 관리해도 좋을 것 같습니다!

Guboneui commented on SaveEarth/Projects/SaveEarth/SaveEarth/Sources/Setting/NotificationSetting/NotificationSettingView.swift:
```diff
@@ -0,0 +1,114 @@
+
+            Spacer()
+        }
+
```
```suggestion
```

Guboneui reviewed (commented):
추가적으로 간단한 리뷰 남겼습니다 ~

importantsgit committed:
STYLE: 리뷰사항 적용

수정 사항
- Date+Extension > Util 내 그룹 생성 후 파일 생성
- 기존 fixed된 width 제거 후 Spacer() 적용

적용 안된 사항
- 디자인 시스템 적용 X
- 외부에서 주입 X > (https://github.com/letusGo-Hack/Team_2_SaveUs/pull/15#discussion_r1742888555)

Marcorable reviewed (approved):
고생하셨습니다!
